lib/Sema/CSDiagnostics.h

```diff
@@ -54,9 +54,12 @@ class FailureDiagnostic {
   /// failure location, types and declarations deduced by
   /// constraint system, and other auxiliary information.
   ///
+  /// \param asNote In ambiguity cases it's beneficial to
+  /// produce diagnostic as a note instead of an error if possible.
+  ///
   /// \returns true If the problem has been successfully diagnosed
   /// and diagnostic message emitted, false otherwise.
-  virtual bool diagnose() = 0;
+  virtual bool diagnose(bool asNote = false) = 0;
```

Comment:
Threading asNote through everything is a bit messy - is there another way you can track it?

Reply:
Originally I wanted to have it as `diagnoseAsNote()` but I'm worried that it could result in a bunch of duplicate code because most likely all the same logic required to produce an error would be required to produce a note...